include/xtypes/idl/IDLParser.hpp

```diff
@@ -0,0 +1,509 @@
+            {
+                scope->fill_all_types(map);
+            }
+        }
+
+        //std::map<std::string, std::shared_ptr<Constants>> constants;
+        //std::map<std::string, std::shared_ptr<Module>> modules;
+        //std::map<std::string, DynamicType::Ptr> types;
+        //std::map<std::string, std::shared_ptr<StructType>> structs;
+        std::map<std::string, DynamicType::Ptr> structs;
+        //std::map<std::string, std::shared_ptr<AnnotationType>> annotations;
+        std::shared_ptr<SymbolScope> outer;
+        std::vector<std::shared_ptr<SymbolScope>> inner;
+    };
+
+    void build_on_ast(
+            const std::shared_ptr<peg::Ast> ast,
+            std::shared_ptr<SymbolScope> scope = nullptr)
+    {
+        using namespace peg::udl;
+        if (scope == nullptr)
+        {
+            types_map_.clear();
+            scope = std::make_shared<SymbolScope>(nullptr);
+        }
+        switch (ast->tag){
+            //case "MODULE_DCL"_:
+            //    module_dcl(ast, scope);
+            //    break;
+            //case "CONST_DCL"_:
+            //    const_dcl(ast, scope);
+            //    break;
+            case "STRUCT_DEF"_:
+                struct_def(ast, scope);
+                break;
+            //case "ANNOTATION_DCL"_:
+            //    annotation_dcl(ast, scope);
+            //    break;
+            default:
+                for (auto node : ast->nodes)
+                {
+                    build_on_ast(node, scope);
+                }
+                break;
+        }
+        scope->fill_all_types(types_map_);
+    }
+
+    void struct_def(
+            const std::shared_ptr<peg::Ast> ast,
+            std::shared_ptr<SymbolScope> outer)
+    {
+        using namespace peg::udl;
+        auto scope = std::make_shared<SymbolScope>(outer);
+        outer->inner.push_back(scope);
+        std::string name;
+        std::map<std::string, Member> member_list;
+        for (const auto& node : ast->nodes)
+        {
+            switch (node->tag)
+            {
+                case "IDENTIFIER"_:
+                {
+                    StructType result(name);
+                    name = node->token;
+                    outer->structs.emplace(name, std::move(result));
+                    break;
+                }
+                case "INHERITANCE"_:
+                    // parent = outer.structs[node->token]; // TODO Check if it doesn't exists
+                    break;
+                case "MEMBER"_:
+                    member_def(node, scope, member_list);
+                    break;
+            }
+        }
+
+        DynamicType::Ptr result = outer->get_type(name);
+        StructType* struct_type = static_cast<StructType*>(const_cast<DynamicType*>(result.get()));
+        for (auto& member : member_list)
+        {
+            struct_type->add_member(std::move(member.second));
+        }
+        // Replace
```

Comment:
To enable forward declarations and recursive structures, I opted by creating the structs once we know the name, and then retrieve it to add its members. As `DynamicType::Ptr` will copy the type, we need to replace the entry in the `structs` map because it isn't the same instance anymore.